sbncode/TPCReco/VertexStub/VertexStubTracker_module.cc

```diff
@@ -181,95 +120,71 @@ void sbn::VertexStubTracker::produce(art::Event& e)
-        thisSlicePFPHits.emplace_back();
-        const std::vector<art::Ptr<recob::Cluster>> &thisPFPClusters = pfparticleClusters.at(i_slc_pfp);
-        art::FindManyP<recob::Hit> clusterHits(thisPFPClusters, e, fPFPLabel);
-        for (unsigned i_clus = 0; i_clus < thisPFPClusters.size(); i_clus++) {
-          thisSlicePFPHits.back().insert(thisSlicePFPHits.back().end(), clusterHits.at(i_clus).begin(), clusterHits.at(i_clus).end());
-        }
-      }
-      slicePFPHits[thisSlice.key()] = thisSlicePFPHits;
-    }
+      sbn::StubInfo sinfo;
+      sinfo.stub = fStubBuilder.FromVertexHit(thisSlice, thisVHit, thisVHitHit, vertex, geo, sce, clock_data, dprop, sinfo.hits, sinfo.pfp); 
+      sinfo.vhit = vhits[i_vhit];
+      sinfo.vhit_hit = vhitHits.at(i_vhit).at(0);
 
-    const std::vector<std::vector<art::Ptr<recob::Hit>>> &thisSlicePFPHits = slicePFPHits.at(thisSlice.key());
+      stubs.push_back(sinfo);
 
-    const recob::Vertex &vert = *vhitVtxs.at(i_vhit).at(0);
-    TVector3 vert_v(vert.position().X(), vert.position().Y(), vert.position().Z());
+    } // end iterate over vertex hits
 
-    std::vector<art::Ptr<recob::Hit>> thisHits = CollectHits(hits, thisVHit, thisVHitHit, vert.position(), geo, dprop);
-
-    // Find which pfparticle to group this VertexStub with 
-    art::Ptr<recob::PFParticle> thisStubPFP = FindPFP(thisHits, thisSlicePFPs, thisSlicePFPHits);
-
-    float charge = 0.;
-    std::set<int> wires;
-    for (unsigned i_hit = 0; i_hit < thisHits.size(); i_hit++) {
-      const recob::Hit &hit = *thisHits[i_hit];
-      charge += fCaloAlg.ElectronsFromADCArea(hit.Integral(), hit.WireID().Plane) * fCaloAlg.LifetimeCorrection(clock_data, dprop, hit.PeakTime(), 0.);
-      wires.insert(hit.WireID().Wire);
+    // Run all of the merging tools
+    for (unsigned i_mrg = 0; i_mrg < fStubMergeTools.size(); i_mrg++) {
+      stubs = fStubMergeTools[i_mrg]->Merge(stubs, geo, sce, clock_data, dprop);
     }
 
-    float vert_w = geo->WireCoordinate(vert.position(), thisVHitHit.WireID());
-    float hit_w = thisVHitHit.WireID().Wire;
-    float wire_dist = abs(hit_w - vert_w);
-
-    sbn::Stub stub;
-    stub.charge = charge;
-    stub.vtx = vert_v;
-    stub.end = thisVHit.spXYZ;
-    stub.pitch = thisVHit.charge / thisVHit.dqdx;
-    stub.nwire = wires.size();
-    stub.wire_dist = wire_dist;
-    stub.plane = thisVHit.wire;
-
     // Save!
-    outStubs->push_back(stub);
-    art::Ptr<sbn::Stub> outStub = StubPtrMaker(outStubs->size() - 1);
-    assn->addSingle(vhits[i_vhit], outStub);
-    for (unsigned i_hit = 0; i_hit < thisHits.size(); i_hit++) { 
-      hitAssn->addSingle(outStub, thisHits[i_hit]);
-    }
-    slcAssn->addSingle(outStub, thisSlice);
-    if (thisStubPFP) {
-      pfpAssn->addSingle(outStub, thisStubPFP);
-    }
-
-  } // end iterate over vertex hits
+    for (unsigned i_stub = 0; i_stub < stubs.size(); i_stub++) {
+      const sbn::StubInfo &sinfo = stubs[i_stub];
```

Comment:
Yeah, I really like range-for loops:
```suggestion
    for (const sbn::StubInfo &sinfo: stubs) {
```